chain/rust/src/assets.rs

```diff
@@ -0,0 +1,595 @@
+                            Some(new_lhs_amount) => {
+                                if new_lhs_amount.is_zero() {
+                                    lhs_assets.remove(asset_name);
+                                    if lhs_assets.is_empty() {
+                                        bundle.remove(policy);
+                                    }
+                                } else {
+                                    *lhs_amount = new_lhs_amount;
+                                }
+                            },
+                            None => {
+                                if T::min_value().is_zero() {
+                                    // if underflow and unsigned, remove
+                                    lhs_assets.remove(asset_name);
+                                    if lhs_assets.is_empty() {
+                                        bundle.remove(policy);
+                                    }
+                                } else {
+                                    // otherwise keep it as minimum
+                                    *lhs_amount = T::min_value();
+                                }
+                            }
+                        },
+                        None => {
+                            // asset name is missing from left hand side
+                            if !T::min_value().is_zero() {
+                                bundle.set(policy.clone(), asset_name.clone(), T::zero().clamped_sub(rhs_amount));
+                            }
+                        }
+                    },
+                    None => {
+                        // policy id missing from left hand side
+                        if !T::min_value().is_zero() {
+                            bundle.set(policy.clone(), asset_name.clone(), T::zero().clamped_sub(rhs_amount));
+                        }
+                    }
+                }
+            }
+        }
+        bundle
+    }
+}
+
+impl<T> AssetBundle<T>
+    where T: num::CheckedAdd + num::CheckedSub + num::Zero + num::Bounded + Copy + Clone,
+{
+    /// Set the value of policy_id:asset_name to value.
+    /// Returns the previous value, or None if it didn't exist
+    pub fn set(&mut self, policy_id: PolicyId, asset_name: AssetName, value: T) -> Option<T> {
+        self.0.entry(policy_id).or_default().insert(asset_name, value)
+    }
+
+    /// Get the value of policy_id:asset_name. If it doesn't exist, 0 is returned.
+    pub fn get(&self, policy_id: &PolicyId, asset_name: &AssetName) -> T {
+        self
+            .0
+            .get(policy_id)
+            .and_then(|assets| assets.get(asset_name))
+            .copied()
+            .unwrap_or(T::zero())
+    }
+
+    /// Adds to bundles together, checking value bounds.
+    /// If an overflow would occur, returns None instead.
+    /// Does not modify self, and instead returns the result.
+    pub fn checked_add(&self, rhs: &Self) -> Option<Self> {
```

Comment:
i think for the calculations it makes sense to add some tests (checked_add, checked_sub, and clamped versions)